libflux/src/semantic/parser/mod.rs

```diff
@@ -0,0 +1,1729 @@
+        tokens: Vec::new(),
+        current_string: String::new(),
+        };
+    lexer.lex_tokens(); 
+    lexer.tokens
+}
+
+// StateFn maintains the lexers recursive state  
+struct StateFn{ f: fn(&mut Lexer) -> Option<StateFn> }
+
+impl Lexer<'_> {
+    // lex_tokens recursively calls the lexer while there are still characters to lex
+    fn lex_tokens(&mut self){
+        let mut state_fn = Some(StateFn{ f: lex_token }); // assigning initial state
+
+        while let Some(func) = state_fn {
+            state_fn = (func.f)(self)
+        }
+    }
+
+    // next grabs the next character from the string while there are still characters to lex
+    fn next(&mut self) -> Option<char> {        
+        match self.source.next() {
+            None => None, 
+            Some(letter) => {
+                if letter.is_alphanumeric() {
+                    self.current_string.push(letter);
+                }
+                Some(letter)
+            }
+        }
+    }
+
+    // emit instatiates a Token and pushes that token along with its TokenType and text, if applicable,
+    // into the Lexer's tokens vector
+    fn emit(&mut self, token: TokenTypes) {
+        if !self.current_string.is_empty() {
+            let text = self.current_string.clone();
+            self.tokens.push(Token{ token_type: token, text: Some(text) })
+        } else {
+            self.tokens.push(Token{ token_type: token, text: None })
+        }
+        self.current_string = String::new();
+    }
+
+    fn keyword_or_ident(&mut self, keyword_map: HashMap<&str, TokenTypes>) {
+        let mut keyword_match = false; 
+        loop {
+            if let Some(letter) = self.source.peek() {
+                if !letter.is_alphanumeric() {
+                    for (keyword, token_type) in &keyword_map {
+                        if keyword == &self.current_string {
+                            // current token is keyword
+                            self.emit(*token_type);
+                            return;
+                        }
+                    }
+                    if !keyword_match{
+                        // token is not a keyword, must be an identifier
+                        self.emit(TokenTypes::IDENTIFIER);
+                        return;
+                    }
+                }
+                self.next();
+            } else {
+                for (keyword, token_type) in &keyword_map {
+                    if keyword == &self.current_string {
+                        // current token is keyword
+                        self.emit(*token_type);
+                        keyword_match = true;
+                        break;
+                    }
+                }
+                if !keyword_match{
+                    // token is not a keyword, must be an identifier
+                    self.emit(TokenTypes::IDENTIFIER);
+                    break;
+                }
+                break;
+            }
+        }
+    }
```

Comment:
The logic between these two blocks in the `if {} else {}` are the same, can you simplify it to not repeat  the code? Maybe create a helper function or change the flow of this code so that its the final step of the function?

Reply:
Thanks for catching that. I meant to go back and refactor this before opening my PR.

Reply:
I think that this update should be a little more DRY.